src/util.rs

```diff
@@ -242,3 +244,67 @@ mod tests {
         assert_eq!(canonicalize_path("foo/bar").unwrap(), "foo/bar");
     }
 }
+
+pub(crate) struct ProgressReader<R, F: Fn(ProgressReaderUpdate)> {
```

Comment:
i assume you looked at `crate::progress::ProgressEmitter` and didn't find it suitable?